### DISCUSSION TITLE
Designer OOP Debugging

### DISCUSSION THREAD

**Discussion by amaitland** [dotnet/winforms#6758] [General]
Do you have any guidance on debugging the designer? 

I get a `The connection to the server has been lost` dialog (be nice to have a meaningful error message here, the `Winforms OOP Designer has terminated unexpectedly` or something meaningful).

There's no information in the dialog, no stack trace. No errors in the event viewer.

Basic steps to reproduce:
- Open form in designer
- Drag a `ChromiumWebBrowser` instance from the toolbox
- Click save button
- Crash occurs.

https://github.com/cefsharp/CefSharp/tree/winforms/oopdesigner

(If someone would actually like to debug this then please let me know and I'll provide more detail)

**Background**

[CefSharp](https://github.com/cefsharp/CefSharp) project provides a `WinForms` implementation with a ultra minimalist designer (basically so you can add the control via the toolbox). `CefSharp` is build using `VC++` and has many unmangaed dependencies. For the classic Designer I've always avoided making any calls that would attempt to load the `VC++ dll`.

Basically I need a stack trace to know what caused the exception.

**Comment by RussKie** [comment]
The OOP designer is a closed source, and while technically you can attach a debugger to the DesignToolsServer.exe you unlikely get far. 
I am currently working on a diagnostics initiative but it will go only so far. Please check this guide https://github.com/dotnet/winforms/blob/main/docs/designer/troubleshooting.md.

`The connection to the server has been lost` is coming from the client when the server unexpectedly stops (e.g., crashes). If the logs don't provide any meaningful error (which can be fixed on your end), then please submit a VS Feedback issue and we'll investigate. Please post a link here, if you do.

If you're trying to get your control to work in the OOP designer, then you'll need to use https://www.nuget.org/packages/Microsoft.WinForms.Designer.SDK/. @KlausLoeffelmann is currently working on the docs and samples.
Please check out https://devblogs.microsoft.com/dotnet/state-of-the-windows-forms-designer-for-net-applications/ as well.

**Reply by amaitland** [reply]
Thanks for the prompt reply, greatly appreciated!

> I am currently working on a diagnostics initiative but it will go only so far. Please check this guide https://github.com/dotnet/winforms/blob/main/docs/designer/troubleshooting.md.

I cranked the logging upto `Verbose` and unfortunately nothing that gives me a clue as to the root cause of the exception.

```
[13:26:48.8575028] Sending request: Objects/RemoveObjectRoots
[13:26:48.8590090] Objects/RemoveObjectRoots took 00:00:00.0015782.
[13:26:49.4072717] Begin PerformFlush()
[13:26:49.4097715] Sending request: Sessions/GetUpdatedRootComponentCodeDom
[13:26:49.5291860] StreamJsonRpc.ConnectionLostException: The JSON-RPC connection with the remote party was lost before the request could complete.
                      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
                      at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__154.MoveNext()
                   --- End of stack trace from previous location where exception was thrown ---
                      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
                      at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__143`1.MoveNext()
                   --- End of stack trace from previous location where exception was thrown ---
                      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
                      at Microsoft.DotNet.DesignTools.Client.DesignToolsClient.<SendRequestAsync>d__49`1.MoveNext()
[13:26:49.5726890] [CefSharp.WinForms.Example.netcore]: info: Microsoft.DotNet.DesignTools.Server.DesignToolsServer[0]
[13:26:49.5726890] [CefSharp.WinForms.Example.netcore]:       Disposing DesignToolsServer [thread 21]
[13:26:49.5840120] End PerformFlush(). Took 00:00:00.1765630.
[13:26:49.6510117] Cannot update selection in server process as Session is not connected.
```

> `The connection to the server has been lost` is coming from the client when the server unexpectedly stops (e.g., crashes)

Didn't take me long to realise that was what had happened. I'd still advocate for a meaningful error message to be added.

From a usability point of view it would be great if the stack trace for the exception that occurred in the `design` process could be populated in the designer error within visual studio. `There is no stack trace or error line information available for this error. ` is all I get currently.

![winforms-oop-designer](https://user-images.githubusercontent.com/307872/156099793-ca6ca0c3-8a48-4e58-8a9e-5ffd346f7aec.PNG)

> If the logs don't provide any meaningful error (which can be fixed on your end), then please submit a VS Feedback issue and we'll investigate.

Unfortunately they don't show anything meaningful as yet. I'll hold off reporting the issue until the official docs have been created, it's possible I'm missing something. 

> If you're trying to get your control to work in the OOP designer, then you'll need to use https://www.nuget.org/packages/Microsoft.WinForms.Designer.SDK/

Thanks, I switched to using that package before posting here in the hopes that it would fix the error. It works until I hit save then the designer crashes.

**Reply by amaitland** [reply]
For what it's worth there's also an exception when I chose not to save the form after it crashes.

```
[13:26:49.7640085] [CefSharp.WinForms.Example.netcore]: Disposing ServerProcess.
[13:43:59.9835203] Cannot update selection in server process as Session is not connected.
[13:44:00.0054020] Exception occurred while disposing VSDesignSurface
                   System.ObjectDisposedException: Cannot access a disposed object.
                   Object name: 'IServiceContainer'.
                      at System.ComponentModel.Design.DesignerHost.System.ComponentModel.Design.IServiceContainer.RemoveService(Type serviceType)
                      at Microsoft.WinForms.DesignTools.Client.Designers.ToolStrips.ClientToolStripService`1.Dispose(Boolean disposing)
                      at Microsoft.WinForms.Utilities.Shared.DisposableObject.DisposeIfNeeded(Boolean disposing)
                      at Microsoft.WinForms.Utilities.Shared.DisposableObject.Dispose()
                      at System.ComponentModel.Design.ServiceContainer.Dispose(Boolean disposing)
                      at System.ComponentModel.Design.DesignSurface.Dispose(Boolean disposing)
                      at Microsoft.VisualStudio.WinForms.VSDesignSurface.Dispose(Boolean disposing)
```

Quick search didn't show an open issue in your issue tracker.

**Reply by RussKie** [reply]
> Didn't take me long to realise that was what had happened. I'd still advocate for a meaningful error message to be added.
> 
> From a usability point of view it would be great if the stack trace for the exception that occurred in the `design` process could be populated in the designer error within visual studio. `There is no stack trace or error line information available for this error. ` is all I get currently.

Thank you for the feedback, I agree the WSOD page needs some attention. I'll add your suggestion to our backlog.

[EDIT] Raised in https://github.com/dotnet/winforms-designer/issues/4156 (internal only).

**Reply by RussKie** [reply]
> For what it's worth there's also an exception when I chose not to save the form after it crashes.
> 
> ```
> [13:26:49.7640085] [CefSharp.WinForms.Example.netcore]: Disposing ServerProcess.
> [13:43:59.9835203] Cannot update selection in server process as Session is not connected.
> [13:44:00.0054020] Exception occurred while disposing VSDesignSurface
>                    System.ObjectDisposedException: Cannot access a disposed object.
>                    Object name: 'IServiceContainer'.
>                       at System.ComponentModel.Design.DesignerHost.System.ComponentModel.Design.IServiceContainer.RemoveService(Type serviceType)
>                       at Microsoft.WinForms.DesignTools.Client.Designers.ToolStrips.ClientToolStripService`1.Dispose(Boolean disposing)
>                       at Microsoft.WinForms.Utilities.Shared.DisposableObject.DisposeIfNeeded(Boolean disposing)
>                       at Microsoft.WinForms.Utilities.Shared.DisposableObject.Dispose()
>                       at System.ComponentModel.Design.ServiceContainer.Dispose(Boolean disposing)
>                       at System.ComponentModel.Design.DesignSurface.Dispose(Boolean disposing)
>                       at Microsoft.VisualStudio.WinForms.VSDesignSurface.Dispose(Boolean disposing)
> ```

Great, this definitely helps to establish the context.
This types of issues need to be raised via VS Feedback, which collects additional information (such as screenshots, memory dumps, stack traces, assembly versions, etc.) that can help us to investigate the issue. Please post the link to the VS Feedback issue here, so that our team can track it.

> 
> Quick search didn't show an open issue in your issue tracker.

This repo is only for the runtime aspects of Windows Forms, i.e. the code that runs a Windows Forms app. The designer repo, where we track issues, is non-public.

**Reply by RussKie** [reply]
I should have asked before, what version of VS are you using?

**Reply by amaitland** [reply]
```
Microsoft Visual Studio Community 2022 (64-bit) - Current
Version 17.1.0
```
